File: app/src/models/cloning-repository.ts
import * as Path from 'path'

let CloningRepositoryID = 1

/** A repository which is currently being cloned. */
export class CloningRepository {
  public readonly id = CloningRepositoryID++
  public readonly path: string
  public readonly url: string

  public constructor(path: string, url: string) {
    this.path = path
    this.url = url
  }

  public get name(): string {
    return Path.basename(this.url, '.git')
  }

  /**
   * A hash of the properties of the object.
   *
   * Objects with the same hash are guaranteed to be structurally equal.
   */
  public get hash(): string {
    return `${this.id}+${this.path}+${this.url}`
  }
}

File: GitDesktop/GitDesktop/Models/CloningRepository.swift
import Foundation

/// A repository currently being cloned. Port of `models/cloning-repository.ts`.
public struct CloningRepository: Sendable, Equatable, Identifiable {
    private static var nextID: Int = 1_000_000

    public var id: Int
    public var path: String
    public var url: String

    public init(path: String, url: String) {
        self.id = CloningRepository.nextID
        CloningRepository.nextID += 1
        self.path = path
        self.url = url
    }

    public var name: String {
        var base = (url as NSString).lastPathComponent
        if base.hasSuffix(".git") { base = String(base.dropLast(4)) }
        return base
    }

    public var hash: String { "\(id)+\(path)+\(url)" }
}
